### PR TITLE
Single route origin simulation

### PR DESCRIPTION
**Pull request by boonious**
For example, the following recipe is applicable to `/news` route.

```
{
    "origin": "https://www.bbc.co.uk/news",
    "route": "/news",
    "stages": [
        {
            "at": 0,
            "latency": "200ms",
            "status": 200
        },
        {
            "at": "10s",
            "latency": "500ms",
            "status": 500
        },
        {
            "at": "30s",
            "latency": "200ms",
            "status": 200
        }
    ]
}
```

**Without route**
Recipe without specified route is applicable to "/*" (default route). This provides the same previous simulation behaviour.

**Wildcard**

Wildcard route examples:
 
```
{  
      "route":"/*",
      "body":"I am ok",
      "stages":[  
         { "at":0, "status":200, "latency":"0ms"},
      ]
}
``` 
Single recipe in a list which is being used for the on-going multi-route simulation dev.
```
[{  
      "route":"/*",
      "body":"I am ok",
      "stages":[  
         { "at":0, "status":200, "latency":"0ms"},
      ]
}]
```

Other wildcard routes, e.g.:

```
{  
      "route":"/news*",
      "body":"I am ok",
      "stages":[  
         { "at":0, "status":200, "latency":"0ms"},
      ]
}
```

### CODE REVIEW

File: lib/origin_simulator.ex
Comment (samfrench) [inline lib/origin_simulator.ex:61]: Do we also want to change the implementation for post to be similar to the behaviour in the get block above?

Comment (boonious) [reply lib/origin_simulator.ex:61]: Can do. Looking at the tests for `POST` (2 _cf._ many GET tests), more tests are also needed for get it working exactly the same way as `GET`. Do we load test with `POST`? 

Will need to change `serve_payload?` too, as per Elixir naming convention (not returning boolean). I didn't follow the convention strictly since the function name makes sense (`do_something?` and not `adjective?` such as `even?`) and Elixir does make exception for those `is_` type boolean functions..

File: lib/origin_simulator/payload.ex
Comment (samfrench) [inline lib/origin_simulator/payload.ex:29]: If we run `mix format` does this get aligned with the other statements?

File: lib/origin_simulator/recipe.ex
Comment (samfrench) [inline lib/origin_simulator/recipe.ex:12]: We should raise a ticket for this.

File: lib/origin_simulator/simulation.ex
Comment (samfrench) [inline lib/origin_simulator/simulation.ex:25]: Should we have a function that matches a recipe list with more than 1 and it returns a 406 (or appropriate status) with a message saying currently we have a limit of 1.

I know it's an edge case, but not sure how the app handles a situation with multiple recipes.

File: lib/origin_simulator/payload.ex
Comment (boonious) [reply lib/origin_simulator/payload.ex:29]: Bizarrely, that was an outcome of running `mix format`!

File: test/origin_simulator/origin_simulator_test.exs
Comment (samfrench) [inline test/origin_simulator/origin_simulator_test.exs]: Probably a typo - I assume it's meant to be response headers.

Secondly, parsed content or passed as in passing through?

File: test/origin_simulator/simulation_test.exs
Comment (samfrench) [inline test/origin_simulator/simulation_test.exs]: We don't call recipe on simulation with this one. Maybe it's the wording or my understanding of this.

File: test/origin_simulator/origin_simulator_test.exs
Comment (boonious) [reply test/origin_simulator/origin_simulator_test.exs]: Typo yeah (response). I also wondered about `passed` vs. `parsed`. This term was already there perhaps to refer to parsing the encoded JSON of the `body` recipe field. So probably `parsed`. I think we should just use `body` to be consistent.

File: test/origin_simulator/single_route_origin_simulator_test.exs
Comment (samfrench) [inline test/origin_simulator/single_route_origin_simulator_test.exs]: We're unlikely to have two products combined, but I see it's an example of an arbitrary route.

Comment (boonious) [reply test/origin_simulator/single_route_origin_simulator_test.exs]: Using path of real news articles now.

File: test/origin_simulator/simulation_test.exs
Comment (boonious) [reply test/origin_simulator/simulation_test.exs]: Fixed, it should be `route()`.

File: test/origin_simulator/origin_simulator_test.exs
Comment (boonious) [reply test/origin_simulator/origin_simulator_test.exs]: Using `parsed body content` now.

File: lib/origin_simulator/simulation.ex
Comment (boonious) [reply lib/origin_simulator/simulation.ex:25]: Posting multiple recipes in a list, gets `406 Not Acceptable` for the time being.

File: lib/origin_simulator.ex
Comment (boonious) [reply lib/origin_simulator.ex:61]: Replicated the GET single-route origin simulation implementation for POST as well.

File: lib/origin_simulator.ex
Comment (ettomatic) [inline lib/origin_simulator.ex]: minor - we could avoid this temporary variable with
```
serve_payload_for_route(conn, Simulation.route(:simulation), conn.request_path)
```
Still quite readable in my opinion

Comment (boonious) [reply lib/origin_simulator.ex]: 👍 shall give the codebase a scan, to see if there are other temporary variables to remove.

**Comment by samfrench** [comment]
Do we need to update the examples and documentation in the readme?

**Comment by boonious** [comment]
> Do we need to update the examples and documentation in the readme?

Yup, immediately after the PRs have been reviewed and merged.

**Review by samfrench — APPROVED** [review]
I approve as it looks good to me. Probably worth getting one or two other approvals as others will have more Elixir knowledge than me.

**Review by woodyblah — APPROVED** [review]

File: lib/origin_simulator.ex
Comment (astalker) [inline lib/origin_simulator.ex:76]: We probably don't need the comment here, it's clear enough I think.

File: lib/origin_simulator/simulation.ex
Comment (astalker) [reply lib/origin_simulator/simulation.ex:25]: "The requested resource is only capable of generating content not acceptable according to the Accept headers sent in the request."

The 406 doesn't seem quite right based on the description for it. Maybe just a 400 bad request would be best

File: lib/origin_simulator/simulation.ex
Comment (astalker) [inline lib/origin_simulator/simulation.ex:29]: A cleaner approach would be to just avoid the comment and add a github comment here (if there is an incoming PR). Comments can so easily go stale or get missed etc. It's this PR yeah https://github.com/bbc/origin_simulator/pull/22

File: test/origin_simulator/multi_route_origin_simulator_test.exs
Comment (astalker) [inline test/origin_simulator/multi_route_origin_simulator_test.exs:10]: I'm not as familiar with Elixir conventions but often I've seen the describe blocks detailing the class/module. So in this case something like

```
describe "MultiRouteOriginSimulator" do
    test "GET pages for a wildcard route should return 200 OK" do
```

File: test/origin_simulator/multi_route_origin_simulator_test.exs
Comment (astalker) [inline test/origin_simulator/multi_route_origin_simulator_test.exs:37]: Would it be better to split these two sets of asserts into two tests? They do look very similar but I'm wondering if there is a reason why we have two sets? If its arbitrary then one set would be ok I think and less ambiguous.

File: test/origin_simulator/origin_simulator_test.exs
Comment (astalker) [inline test/origin_simulator/origin_simulator_test.exs:62]: Should be ok to remove the comment here, the test description is clear enough

File: test/origin_simulator/origin_simulator_test.exs
Comment (astalker) [inline test/origin_simulator/origin_simulator_test.exs:85]: I think we're missing some detail in some of these test descriptions (not necessarily introduced in this PR of course). In this one for example, the description doesn't really us why there is random latency.

File: lib/origin_simulator/simulation.ex
Comment (boonious) [reply lib/origin_simulator/simulation.ex:25]: I agree since there's no accept headers involved. The 406 has always been there as the standard response. Let's create a task to fix this later.

File: lib/origin_simulator.ex
Comment (boonious) [reply lib/origin_simulator.ex:76]: These functions no longer exist (next PR). Since the root wildcard `/*` becomes default route with a standard response, every request that matches `get "/*glob"` will now be served.

File: test/origin_simulator/multi_route_origin_simulator_test.exs
Comment (boonious) [reply test/origin_simulator/multi_route_origin_simulator_test.exs:37]: To demonstrate that `/*` works for any-many routes. Two paths should be the minimum testing criteria. This should be refactored in the same way as the single-route tests (from the next PR), cf. the following "arbitrary wildcard" test case :

```ex
    test "for '/*' route", %{payload: payload} do
      payload = Map.put(payload, "route", "/*")
      conn(:post, "/#{admin_domain()}/add_recipe", Poison.encode!(payload)) |> OriginSimulator.call([])
      Process.sleep(20)

      conn(:get, "/news/uk-politics-51287430")
      |> OriginSimulator.call([])
      |> assert_status_body(200, body_mock())

      conn(:get, "/sport/tennis/51291122")
      |> OriginSimulator.call([])
      |> assert_status_body(200, body_mock())
    end

    test "for arbitrary wildcard route", %{payload: payload} do
      payload = Map.put(payload, "route", "/news*")
      conn(:post, "/#{admin_domain()}/add_recipe", Poison.encode!(payload)) |> OriginSimulator.call([])
      Process.sleep(20)

      conn(:get, "/news/uk-politics-51287430")
      |> OriginSimulator.call([])
      |> assert_status_body(200, body_mock())
    end
```

Actually need to improve the test coverage for multi-route simulation!

File: lib/origin_simulator.ex
Comment (astalker) [reply lib/origin_simulator.ex:76]: Yep, that's cool. It's more of a general thing regarding comments as we try to avoid them as much as possible.

File: test/origin_simulator/multi_route_origin_simulator_test.exs
Comment (astalker) [reply test/origin_simulator/multi_route_origin_simulator_test.exs:37]: I guess you could say that two is still arbitrary as there is no guarantee the code isn't hardcoded to handle the two paths. I think in the original code here that testing the route `/any_path` or `/some-wildcard-path` would be better than `/news/uk-politics-51287430` as that helps describe the intent of it being a wildcard path.

**Comment by astalker** [comment]
> > Do we need to update the examples and documentation in the readme?
> 
> Yup, immediately after the PRs have been reviewed and merged.

I don't want to hold things up of course but it's often a good practice to include documentation as part of the PR. That helps ensure it doesn't get missed and also means that if changes are reverted they go hand in hand. Git log history is subjective but personally I like to be able to go back in time to see the changes and view any docs changes alongside.

File: test/origin_simulator/multi_route_origin_simulator_test.exs
Comment (boonious) [reply test/origin_simulator/multi_route_origin_simulator_test.exs:10]: Thanks @astalker, I made the description changes and reorganisation for the tests according to our chat, adopted a little bit of the cucumber format. Committed the change in https://github.com/bbc/origin_simulator/pull/22

**Review by astalker — APPROVED** [review]

**Review by ettomatic — APPROVED** [review]

**Pull request merged**
